### PR TITLE
Cache

### LINKED ISSUE #37: Caching
We should set up our server methods that read from the DB to use caching where appropriate.

### PR DESCRIPTION
**Pull request by cadecairos**
Fixes #37

### CODE REVIEW

File: services/api/lib/queries.js [hunk 1]
```diff
@@ -83,9 +83,9 @@ module.exports = {
     "projects.deleted_at is NULL AND projects.deleted_at IS NULL ORDER BY created_at DESC LIMIT $1 OFFSET $2",
 
     // Find projects created by given user
-    // Params: user_id bigint, limit integer, offset integer
+    // Params: user_id bigint, offset integer
     findUsersProjects: "SELECT " + projectCols + " FROM projects INNER JOIN users ON users.id = projects.user_id " +
-    "WHERE projects.user_id = $1 AND projects.deleted_at IS NULL ORDER BY created_at DESC LIMIT $2 OFFSET $3",
+    "WHERE projects.user_id = $1 AND projects.deleted_at IS NULL ORDER BY created_at DESC LIMIT 25 OFFSET $2",
```
Comment (thisandagain) [inline services/api/lib/queries.js]: Why the hard limit here?

Comment (cadecairos) [inline services/api/lib/queries.js]: a variable limit makes caching and invalidations practically impossible.

Comment (thisandagain) [inline services/api/lib/queries.js]: Hmm ... generally I include the limit / offset combination in the cache key. Any reason why we couldn't do that here as well?

**Comment by thisandagain** [comment]
@ashleygwilliams R?

File: services/api/lib/queries.js (hunk 1 above)
Comment (cadecairos) [inline services/api/lib/queries.js]: that's simple enough. How do we handle invalidations on project create/remix (forgot to do that, adding now)

should I just invalidate the most used limit (25, by app project view), and just page 1?

Comment (thisandagain) [inline services/api/lib/queries.js]: As we are using redis generally I would recommend you expire a wildcard. If you cache keys look like `get::blah::limit=x::offset=y` you would expire `get::blah::*`.

Comment (cadecairos) [inline services/api/lib/queries.js]: I wonder if that works with catbox-redis/catbox-memory

**Comment by cadecairos** [comment]
I need to do some more work here.

**Comment by cadecairos** [comment]
@toolness I think I've got a fix for this, just need to write regression tests.

**Comment by ashleygwilliams** [comment]
on first run from fresh clone i get a single test failure:

```
Test script errors:

Multiple callbacks or thrown errors received in test "Before " (error): Expected 200 to equal specified value
      at /Users/ag_dubs/Projects/api.webmaker.org/test/services/api/handlers/elements.js:878:40
      at process._tickDomainCallback (node.js:381:11)

There were 1 test script error(s).

Failed tests:

  84) Element Handlers Delete Element success - owner:

      Timed out (5000ms) - Element Handlers Delete Element success - owner



1 of 367 tests failed
Test duration: 6982 ms
Assertions count: 1564 (verbosity: 4.26)
```

**Comment by ashleygwilliams** [comment]
ran again 2x, same issue

File: test/services/api/handlers/pages.js [hunk 2]
```diff
@@ -7,7 +7,8 @@ var configs = require('../../../fixtures/configs/page-handlers'),
   after = lab.after,
   test = lab.test,
   expect = require('code').expect,
-  server;
+  server,
+  thumnailServiceUrl;
```
Comment (ashleygwilliams) [inline test/services/api/handlers/pages.js]: is there are reason for leaving out the 'b' in thumbnail?

Comment (cadecairos) [inline test/services/api/handlers/pages.js]: nah, just a typo. my finger misses the b way too often. good catch

Comment (ashleygwilliams) [inline test/services/api/handlers/pages.js]: you are very consistent with it :)

**Comment by ashleygwilliams** [comment]
having added https://irccloud.mozilla.com/pastebin/n5FiB31v  (and requiring `Hoek`) to `/test/mocks/server.js` we now have more test errors. which really makes no sense, since that addition is just for better logging. here they are:

https://gist.github.com/ashleygwilliams/ed436fd9af1561067f84

**Comment by ashleygwilliams** [comment]
ran a clean build of `develop` to ensure that this issue is local to this PR. confirmed.

**Comment by cadecairos** [comment]
Okay, I figured out whats going on here.

It's a race condition where the previous request runs some tail functions, and emits a tail event on the connection after the failing test case begins, triggering the evbent listener too soon.

If I can find a way to remove event listeners on the server, I can check to make sure the tail is being called on the right request.

**Comment by cadecairos** [comment]
OK, Latest commit should fix any issues with the unit tests.

Giv'r a go @ashleygwilliams

**Comment by ashleygwilliams** [comment]
tests are passing for me, 5/5x. gonna take a look at the code now.

File: services/api/lib/cache.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    });
+  }
+
+  function invalidateKey(type, funcName, keys, done) {
+    if ( !Array.isArray(keys) ) {
+      keys = [keys];
+    }
+
+    process.nextTick(function() {
+      debug({
```
Comment (ashleygwilliams) [inline services/api/lib/cache.js]: did you mean to leave this in here?

Comment (cadecairos) [inline services/api/lib/cache.js]: yup. depending on the LOG_LEVEL the bunyan logger handles what to do with this.

**Comment by ashleygwilliams** [comment]
looks good to me. final comment: perhaps make the expires/stales time consistent?

**Comment by ashleygwilliams** [comment]
r+ :fireworks:

**Pull request merged**